Get single or set of row's activity

TendTo opened the pull request:
## Major

- Added two commodity overloads on the method `computePrimalActivity` to selectively compute only the activity of a single row or selected rows.
- Added `getRowPrimalValue` along with several variations to the solver. The idea is to have a way to efficiently get the value of a row computed with the current primal solution.

## Minor

- Use default destructor definition in `SPxException` to avoid the warning: `warning: definition of implicit copy constructor for 'SPxException' is deprecated because it has a user-declared destructor [-Wdeprecated]`
- Use const reference instead of const copy for the `what()` getter in `SPxException`

TendTo committed:
feat: add single row functions to compute primal activity

TendTo committed:
chore: fix linting error in destructor definition and msg getter

TendTo committed:
fix: add missing include of <set>

TendTo committed:
feat: add getRowPrimalValue method and variants

TendTo committed:
chore: invert parameters order for consistency

TendTo committed:
fix: use [] operator instead of value

TendTo committed:
chore: add default copy constructors to SpxException

TendTo committed:
fix: missing const in copy constructor

leoneifler commented:
Hi, thank you for submitting this PR. I will add some comments to the code. 

Just so you know our current workflow: Once we think the code is good, I will cherry-pick your PR onto our internal GitLab, as that is where all of our CI runs (migrating fully to GitHub is planned, but the CI migration proves difficult). I will then close your PR, and the changes will be mirrored to GitHub (your commits will still be in your name, of course)

TendTo committed:
fix: add other explicit default copy constructors

leoneifler commented on src/soplex.h:
```diff
@@ -663,6 +663,20 @@ class SoPlexBase
+
+   /// Get the value for the row with indices in \p indices given the current primal solution
+   bool getRowsPrimalValue(const std::set<int>& indices, VectorBase<R>& vector);
+   bool getRowsPrimalValueReal(const std::set<int>& indices, R* p_vector, int i);
```
rename the i to something descriptive like size

TendTo replied:
I guess I will use `dim` for consistency

https://github.com/scipopt/soplex/blob/07911134f4c7e478b9131895eeabc15d652ecb29/src/soplex.h#L668

TendTo committed:
fix: use vectors instead of sets

While sets guarantees uniqueness in the entries, that may not be something the user wants. Furthermore, vectors are more usually more efficient when it comes to looping over them, which is the only operation performed in this case.
The user can still utilise a set and then convert it to a vector before passing it to the function.

leoneifler commented on src/soplex.h:
```diff
@@ -663,6 +663,20 @@ class SoPlexBase
+
+   /// Get the value for all rows given the current primal solution
+   bool getRowsPrimalValue(VectorBase<R>& vector);
+   bool getRowsPrimalValueReal(R* p_vector, int i);
```
again, rename `i`

TendTo replied:
Same as above

leoneifler commented on src/soplex.hpp:
```diff
@@ -3860,6 +3860,114 @@ bool SoPlexBase<R>::getPrimalRational(VectorBase<Rational>& vector)
+  if(_rationalLP != 0 && hasSol())
+  {
+    _syncRationalSolution();
+    value = _rationalLP->computePrimalActivity(i, _solRational._primal);
```
The `unscaled` flag always defaults to `true`, right? So why do you set it sometimes, but not all the time?

TendTo replied:
This confused me too, but there are a couple of instances when dealing with `_realLP` that true is used explicitly, while in most cases, it is not. I had no idea if this difference had any semantic meaning, so I kept it. 

E.g.
https://github.com/scipopt/soplex/blob/3982032d3aa38d37697b82ffd9c95cf225b772ce/src/soplex.hpp#L1267

If you say it doesn't, it would be best to remove it

leoneifler replied:
While I did not write this code myself, I think it is safe to remove it.

leoneifler commented on src/soplex.hpp:
```diff
@@ -3860,6 +3860,114 @@ bool SoPlexBase<R>::getPrimalRational(VectorBase<Rational>& vector)
+bool SoPlexBase<R>::getRowsPrimalValueReal(const std::set<int>& indices, R* p_vector, int i) {
+  if(_realLP != 0 && hasSol())
+  {
+    assert(i >= static_cast<int>(indices.size()));
```
does the `i` even make sense here? Why not just use `indices.size()`?

TendTo replied:
Since `i` is really the size of the `R` `p_vector`, this assertion ensures that there are at least `indices.size()` slots in `p_vector`. Changing `i` to `dim` will help with clarity.
This said, `dim` should really be `>= nRows()` instead of `indices.size()`, unless you want to deal with sparse vectors in this context.

TendTo committed:
fix: correct behaviour in computePrimalActivity with indexes

TendTo committed:
docs: update docstrings according to the last changes

TendTo commented on src/soplex/spxlpbase_real.hpp:
```diff
@@ -122,6 +122,106 @@ void SPxLPBase<R>::computePrimalActivity(const VectorBase<R>& primal, VectorBase
+      if(unscaled && _isScaled)
+      {
+        lp_scaler->getColUnscaled(*this, c, tmp);
+        activity.multAdd(primal[c], tmp);
```
Yep, this is an error. I corrected this for the rational case, but I missed this one

TendTo committed:
refactor: fix naming convention

TendTo committed:
fix: consistent method for rational

leoneifler commented on src/soplex.h:
```diff
@@ -663,6 +663,20 @@ class SoPlexBase
    bool getPrimalRational(VectorRational& vector);
 
+   /// Get the value for the \p i 'th row given the current primal solution
+   bool getRowPrimalValue(int i, R& value);
```
Yeah the naming of variables is not great in the existing code. Then lets stay with `i`. The renaming of the existing code should be done in a separate refactoring branch

leoneifler commented:
Could you add a `CHANGELOG` entry? Otherwise the PR looks good to me.

TendTo committed:
fix: removed unneeded true parameter

TendTo committed:
chore: update CHANGELOG

DominikKamp commented:
@TendTo Finally, a variant is merged with fd7f66bcc268950a403444444f0e70ac847e72c0 including dual counterparts `getColActivity()` and `getColsActivity()`. If more should be added, please open another pull request. Thank you for the contribution and sorry for the delay.
